internal/commands/build/build.go

```diff
@@ -0,0 +1,216 @@
+	flags.BoolVar(&opts.noCache, "no-cache", false, "Do not use cache when building the image")
+	flags.StringVar(&opts.progress, "progress", "auto", "Set type of progress output (auto, plain, tty). Use plain to show container output")
+	flags.BoolVar(&opts.pull, "pull", false, "Always attempt to pull a newer version of the image")
+
+	return cmd
+}
+
+func runBuild(dockerCli command.Cli, application string, opt buildOptions) (reference.Named, error) {
+	err := checkMinimalEngineVersion(dockerCli)
+	if err != nil {
+		return nil, err
+	}
+
+	if opt.tag == "" {
+		// FIXME temporary, until we get support for Digest in bundleStore and other commands
+		return nil, fmt.Errorf("A tag is required to run docker app build")
+	}
+
+	var ref reference.Named
+	ref, err = packager.GetNamedTagged(opt.tag)
+	if err != nil {
+		return nil, err
+	}
+
+	app, err := packager.Extract(application)
+	if err != nil {
+		return nil, err
+	}
+	defer app.Cleanup()
+
+	buildopts, err := parseCompose(app, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	buildopts["com.docker.app.invocation-image"], err = createInvocationImageBuildOptions(dockerCli, app)
+	if err != nil {
+		return nil, err
+	}
+
+	debugBuildOpts(buildopts)
+
+	ctx, cancel := context.WithCancel(appcontext.Context())
+	defer cancel()
+	const drivername = "buildx_buildkit_default"
+	d, err := driver.GetDriver(ctx, drivername, nil, dockerCli.Client(), nil, "", nil)
+	if err != nil {
+		return nil, err
+	}
+	driverInfo := []build.DriverInfo{
+		{
+			Name:   "default",
+			Driver: d,
+		},
+	}
+
+	pw := progress.NewPrinter(ctx, os.Stderr, opt.progress)
+
+	// We rely on buildx "docker" builder integrated in docker engine, so don't need a DockerAPI here
+	resp, err := build.Build(ctx, driverInfo, buildopts, nil, dockerCli.ConfigFile(), pw)
+	if err != nil {
+		return nil, err
+	}
+	fmt.Fprintln(dockerCli.Out(), "Successfully built service images") //nolint:errcheck
+
+	bundle, err := packager.MakeBundleFromApp(dockerCli, app, nil)
+	if err != nil {
+		return nil, err
+	}
+	updateBundle(bundle, resp)
+
+	if ref == nil {
+		if ref, err = computeDigest(bundle); err != nil {
+			return nil, err
+		}
+	}
+
+	if err = packager.PersistInBundleStore(ref, bundle); err != nil {
+		return nil, err
+	}
+	return ref, nil
+}
+
+func checkMinimalEngineVersion(dockerCli command.Cli) error {
+	info, err := dockerCli.Client().Info(appcontext.Context())
+	if err != nil {
+		return err
+	}
+	majorVersion, err := strconv.Atoi(strings.SplitN(info.ServerVersion, ".", 2)[0])
+	if err != nil {
+		return err
+	}
+	if majorVersion < 19 {
+		return errors.New("'build' require docker engine 19.03 or later")
+	}
+	return nil
+}
+
+func updateBundle(bundle *bundle.Bundle, resp map[string]*client.SolveResponse) {
+	debugSolveResponses(resp)
+	for service, r := range resp {
+		digest := r.ExporterResponse["containerimage.digest"]
+		if service == "com.docker.app.invocation-image" {
+			bundle.InvocationImages[0].Digest = digest
```

Comment:
It would be great to get the size of the image too here, along with the digest, so we can set digest and size fields (needed with the PR I'm preparing to handle relocation maps). See https://github.com/docker/cnab-to-oci/blob/master//converter/convert.go#L228

Reply:
Would need a fix in Buildx then